### PR TITLE
Remove cereal submodule

### PR DESCRIPTION
**Pull request by elad335**
* Replaced cereal with c++20 custom alternative.
* Fix RSX capture versioning checks.

### CODE REVIEW

File: rpcs3/util/serialization.hpp [hunk 1]
```diff
@@ -0,0 +1,252 @@
+#pragma once
+
+#include "util/types.hpp"
+#include <vector>
+
+struct serial_load
```
Comment (Megamouse) [inline rpcs3/util/serialization.hpp:6]: Can't you call it serialize and deserialize?

Comment (Megamouse) [reply rpcs3/util/serialization.hpp:6]: Or serializer and deserializer

File: rpcs3/util/serialization.hpp [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+		}
+	}
+
+	explicit operator bool() const
+	{
+		return pos <= data.size();
+	}
+
+	template <typename... T>
+	bool operator()(T&... args)
```
Comment (Megamouse) [inline rpcs3/util/serialization.hpp:164]: I think using this operator is not very intuitive.
Maybe use << and >> respectively

Comment (Megamouse) [reply rpcs3/util/serialization.hpp:164]: Or set and get

Comment (elad335) [reply rpcs3/util/serialization.hpp:164]: Most serialization libraries do it this way and it is very convenient. Furthermore I would like to not change existing code for cereal in RSX captures.

File: rpcs3/util/serialization.hpp (hunk 1 above)
Comment (elad335) [reply rpcs3/util/serialization.hpp]: What about serial and deserial? I just can predict the many typos people will make with full naming.

Comment (elad335) [reply rpcs3/util/serialization.hpp]: ok changed.

File: rpcs3/Emu/RSX/Capture/rsx_replay.h [hunk 3]
```diff
@@ -157,7 +162,27 @@ namespace rsx
 		void serialize(Archive & ar)
 		{
 			ar(magic);
+
+			// Check if desirializing
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/Capture/rsx_replay.h:166]: Typo

File: rpcs3/util/serialization.hpp [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+
+		// std::array, C-array
+		template <typename T>
+		bool read(T& obj) requires array_alike<T>
+		{
+			if constexpr (std::is_trivially_copyable_v<std::remove_reference_t<decltype(std::declval<T>()[0])>>)
+			{
+				return read(std::data(obj), sizeof(obj[0]) * std::size(obj));
+			}
+			else
```
[the remark this answers is not in the document]
Comment (elad335) [reply rpcs3/util/serialization.hpp]: Fixed.

File: rpcs3/Emu/RSX/Capture/rsx_replay.h (hunk 3 above)
Comment (elad335) [reply rpcs3/Emu/RSX/Capture/rsx_replay.h]: Fixed.

File: rpcs3/rpcs3.vcxproj [hunk 5]
```diff
@@ -62,13 +62,12 @@
...
   <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|x64'">
     <LibraryPath>$(SolutionDir)lib\$(Configuration)-$(Platform)\;$(UniversalCRT_LibraryPath_x64);$(LibraryPath)</LibraryPath>
-    <IncludePath>..\3rdparty\7z\src;..\3rdparty\hidapi\hidapi\hidapi;.\;..\;..\3rdparty\asmjit\asmjit\src;..\3rdparty\yaml-cpp\include;..\3rdparty\ffmpeg\include;$(VC_IncludePath);$(WindowsSDK_IncludePath);$(UniversalCRT_IncludePath);..\3rdparty\XAudio2Redist\include;..\3rdparty\libpng\libpng;..\3rdparty\GL;..\3rdparty\stblib\include;..\3rdparty\OpenAL\include;..\3rdparty\pugixml\src;..\3rdparty\Optional;..\3rdparty\discord-rpc\include;..\3rdparty\zlib</IncludePath>
-  </PropertyGroup>
+    <IncludePath>..\3rdparty\7z\src;..\3rdparty\hidapi\hidapi\hidapi;.\;..\;..\3rdparty\asmjit\asmjit\src;..\3rdparty\yaml-cpp\include;..\3rdparty\ffmpeg\include;$(VC_IncludePath);$(WindowsSDK_IncludePath);$(UniversalCRT_IncludePath);..\3rdparty\XAudio2Redist\include;..\3rdparty\libpng\libpng;..\3rdparty\GL;..\3rdparty\stblib\include;..\3rdparty\OpenAL\include;..\3rdparty\pugixml\src;..\3rdparty\Optional;..\3rdparty\discord-rpc\include;..\3rdparty\zlib\zlib</IncludePath>  </PropertyGroup>
```
Comment (Megamouse) [inline rpcs3/rpcs3.vcxproj:70]:
```suggestion
    <IncludePath>..\3rdparty\7z\src;..\3rdparty\hidapi\hidapi\hidapi;.\;..\;..\3rdparty\asmjit\asmjit\src;..\3rdparty\yaml-cpp\include;..\3rdparty\ffmpeg\include;$(VC_IncludePath);$(WindowsSDK_IncludePath);$(UniversalCRT_IncludePath);..\3rdparty\XAudio2Redist\include;..\3rdparty\libpng\libpng;..\3rdparty\GL;..\3rdparty\stblib\include;..\3rdparty\OpenAL\include;..\3rdparty\pugixml\src;..\3rdparty\Optional;..\3rdparty\discord-rpc\include;..\3rdparty\zlib\zlib</IncludePath>
  </PropertyGroup>
```

**Comment by Nekotekina** [comment]
The idea of "cracking down" all std containers and possible some custom ones into few categories based on concepts is pretty powerful, although reckless as of now and needs a lot of refining of concept usage.

**Comment by Nekotekina** [comment]
Made few changes, notably removed explicit bool operator since it conflicts with deserialization operator template.

File: rpcs3/util/serialization.hpp [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+
+				if constexpr (std::is_trivially_copyable_v<typename std::remove_reference_t<T>::value_type>)
+				{
+					raw_serialize(obj.data(), sizeof(obj[0]) * obj.size());
+				}
+				else
+				{
+					for (auto&& value : obj)
+					{
+						serialize(value);
```
Comment (elad335) [inline rpcs3/util/serialization.hpp:122]: Needs an early return if returns false.

File: rpcs3/util/serialization.hpp [hunk 7]
```diff
@@ -0,0 +1,339 @@
...
+				for (usz i = obj.size();;)
+				{
+					const usz i_old = std::exchange(i, i >> 7);
+					operator()<u8>(static_cast<u8>(i_old % 0x80) | (u8{!!i} << 7));
+					if (!i) break;
+				}
+
+				for (auto&& value : obj)
+				{
+					serialize(value);
```
Comment (elad335) [inline rpcs3/util/serialization.hpp:212]: same

File: rpcs3/util/serialization.hpp [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+				return std::bit_cast<type>(buf);
+			}
+			else if constexpr (std::is_constructible_v<type, stx::exact_t<serial&>>)
+			{
+				return type(stx::exact_t<serial&>(*this));
+			}
+			else if constexpr (std::is_constructible_v<type>)
+			{
+				type value{};
+				serialize(value);
```
Comment (elad335) [inline rpcs3/util/serialization.hpp:322]: maybe assert its return value.

**Pull request merged**